retworkx-core/src/traversal/bfs_visit.rs

```diff
@@ -0,0 +1,282 @@
+/// ```
+pub fn breadth_first_search<G, I, F, C>(
+    graph: G,
+    starts: I,
+    mut visitor: F,
+) -> C
+where
+    G: IntoEdges + Visitable,
+    I: IntoIterator<Item = G::NodeId>,
+    F: FnMut(BfsEvent<G::NodeId, &G::EdgeWeight>) -> C,
+    C: ControlFlow,
+{
+    let discovered = &mut graph.visit_map();
+    let finished = &mut graph.visit_map();
+
+    for start in starts {
+        // `bfs_visitor` returns a "signal" to either continue or exit early
+        // but it never "prunes", so we use `unreachable`.
+        try_control!(
+            bfs_visitor(graph, start, &mut visitor, discovered, finished),
+            unreachable!()
+        );
+    }
+    C::continuing()
+}
+
+fn bfs_visitor<G, F, C>(
+    graph: G,
+    u: G::NodeId,
+    visitor: &mut F,
+    discovered: &mut G::Map,
+    finished: &mut G::Map,
+) -> C
+where
+    G: IntoEdges + Visitable,
+    F: FnMut(BfsEvent<G::NodeId, &G::EdgeWeight>) -> C,
+    C: ControlFlow,
+{
+    if !discovered.visit(u) {
+        return C::continuing();
+    }
+
+    try_control!(visitor(BfsEvent::Discover(u)), {}, {
+        let mut stack: VecDeque<G::NodeId> = VecDeque::new();
+        stack.push_front(u);
+
+        while let Some(u) = stack.pop_front() {
+            for edge in graph.edges(u) {
+                let v = edge.target();
+                if !discovered.is_visited(&v) {
+                    try_control!(
+                        visitor(BfsEvent::TreeEdge(u, v, edge.weight())),
+                        continue
+                    );
+                    discovered.visit(v);
+                    try_control!(visitor(BfsEvent::Discover(v)), continue);
+                    stack.push_back(v);
+                } else {
+                    // non - tree edge.
+                    try_control!(
+                        visitor(BfsEvent::NonTreeEdge(u, v, edge.weight())),
+                        continue
+                    );
+
+                    if !finished.is_visited(&v) {
+                        try_control!(
+                            visitor(BfsEvent::GrayTargetEdge(
+                                u,
+                                v,
+                                edge.weight()
+                            )),
+                            continue
+                        );
+                    } else {
+                        try_control!(
+                            visitor(BfsEvent::BlackTargetEdge(
+                                u,
+                                v,
+                                edge.weight()
+                            )),
+                            continue
+                        );
+                    }
+                }
+            }
+
+            let first_finish = finished.visit(u);
+            debug_assert!(first_finish);
+            try_control!(
+                visitor(BfsEvent::Finish(u)),
+                panic!("Pruning on the `BfsEvent::Finish` is not supported!")
```

Comment:
I get why you did this since a panic will raise to the user directly without needing to manually propogate an err return through the whole path. Also in the python case this gets exposed as a `PanicException` with the error message.

But I wonder if we want to have an error return associated with this. Especially for `retworkx-core` users it's a bit off-putting I think because to handle this they need to `catch_unwind` a call to `breadth_first_search()`. Especially as this is a user error right? It's a mistake for the user to request a `Prune` on a `Finish`  event.

Reply:
Yeah, it's a user mistake and I'm in favor of using panic rather returning a `Result` since I consider this as a bug that should be fixed in their code rather giving them the chance to "recover". Also defining an error here would only complicate user valid code, i.e they might have to deal with return types like `Result<Result<_, _>, _>`.

Reply:
Ok, that's fine, we have it documented as panicking in this case so that should be enough